src/compilation/config.rs

```diff
@@ -0,0 +1,128 @@
+use crate::{
+    content_encoding::{
+        ContentTypeCheckType, ContentTypeConverterType, DEFAULT_CONTENT_ENCODING_CHECK,
+        DEFAULT_CONTENT_ENCODING_CONVERT,
+    },
+    media_type::{MediaTypeCheckType, DEFAULT_MEDIA_TYPE_CHECK},
+    schemas,
+};
+use serde_json::Value;
+use std::{collections::HashMap, fmt};
+
+#[allow(missing_docs)]
+#[derive(Clone, Default)]
+pub struct JSONSchemaCompileConfig {
+    draft: Option<schemas::Draft>,
+    custom_content_media_type_checks: HashMap<String, Option<MediaTypeCheckType>>,
+    custom_content_encoding_checks: HashMap<String, Option<ContentTypeCheckType>>,
+    custom_content_encoding_converters: HashMap<String, Option<ContentTypeConverterType>>,
+}
+
+#[allow(missing_docs)]
+impl JSONSchemaCompileConfig {
+    pub(crate) fn draft(&self) -> schemas::Draft {
+        self.draft
+            .expect("JSONSchema::compile should have defined a specific draft version.")
```

Comment:
Not sure if it should be failable at all... e.g. when is it ok to panic? We always have a default - Draft 7 at the moment. Having an infallible data layout will help us to avoid extra actions like `set_draft_if_missing` which seems unnecessary to me

Reply:
That's correct, but this is partially needed to allow users to not specify the Draft version when `JSONSchema` is created (or in `jsonschema::is_valid`).

In compilation phase we would need to allow the user to specify or not specify the draft in the configuration.
If the draft is not configured (set to `None`) then we try to detect it from the schema and/or default it to `Draft7` (this is what `set_draft_if_missing` does).

We might be looking for an easier approach or just avoid the panic here by defaulting to`Draft::default()`.

My personal take is that an `expect` here would be acceptable because the invariance is guaranteed by the library and not by user-code and tests would quickly catch this issue. Using `Draft::default()` here would potentially hide the issue and lead to a "broken" release (where draft7 is used if the user has not specified the draft version)

Reply:
I think that having ``pub(crate) draft: schemas::Draft`` is better, because we statically don't have a state where it is `None` and we don't need tests for it or using `expect` at all. If the user doesn't want to specify a draft, then they can rely on `Default::default()`. 

> Using Draft::default() here would potentially hide the issue and lead to a "broken" release (where draft7 is used if the user has not specified the draft version)

Hm, this is the current behavior, right? I mean it seems expected to me if I don't specify a draft version / config completely, then I let the library select it for me. 

Maybe we can use some kind of builder pattern instead.

```rust
let config = CompilationConfig::options().draft(Draft7).build();  // Or other naming - but some kind of builder pattern
let compiled = JSONSchema::compile(schema, Some(config));
```

Or in the `JSONSchema` itself (like it is in `url::Url` )

```rust
let compiled = JSONSchema::options().draft(Draft7).compile(schema)?;
```

What do you think?

Reply:
> Hm, this is the current behavior, right?

Not quite, the current implementation (either master or this PR) does try to detect the version from the input schema and then if still no draft is defined we use `Draft7`.
Let's use the following schema as example
```json
{
    "$schema": "http://json-schema.org/draft-06/schema#",
    ...
}
```
If we do compile this without defining the draft the schema should be compiled with `Draft6`, but if we do use `Draft::default()` it would use `Draft7`.

So we still need to preserve the information of "was draft defined by the user?"

---

About using the `JSONSchema::options`, it might be a clever idea as it would remove the naming difficulty and adhere to a pattern that seems getting more and more traction in the rust community.

I might try to make some progress on this maybe over the weekend though :)

Reply:
Yep, I meant using `default()` as the last option in that chain (how it is actually done now, but with explicitly using `Default::default()`)

> About using the JSONSchema::options, it might be a clever idea as it would remove the naming difficulty and adhere to a pattern that seems getting more and more traction in the rust community.

> I might try to make some progress on this maybe over the weekend though :)

I think it might be cool to have it as a separate PR to reduce the affected surface :) and then rebase

And, Samuele, thank you very much for working on this! :)

Reply:
> I think it might be cool to have it as a separate PR

I agree ... over the weekend I'll try to create a PR where options are defined and they will have only the `Draft` handling (so it will small and nicer) ... after that other PRs will be spawned to add contentMediaType and contentMediaEncoding.

The idea of having them over here (discussion PR) was to show why I was thinking to enable a similar flow.
Sometimes "seeing" the effect is more effective than a very long and abstract description 😎